nixos/modules/services/web-apps/tachidesk-server.nix

```diff
@@ -0,0 +1,235 @@
+          '';
+        };
+
+        localSourcePath = mkOption {
+          type = types.path;
+          default = cfg.dataDir;
+          defaultText = lib.literalExpression "tachidesk-server.dataDir";
+          example = "/var/data/local_mangas";
+          description = mdDoc ''
+            Path to the local source folder.
+          '';
+        };
+      };
+
+      extraConfig = mkOption {
+        type = types.submodule {
+          freeformType = with lib.types; attrsOf str;
+        };
+        default = { };
+        description = mdDoc ''
+          Extra configuration to write to {file}`server.conf`.
+          See <https://github.com/Suwayomi/Tachidesk-Server/wiki/Configuring-Tachidesk-Server> for more information.
+        '';
+        example = {
+          "server.socksProxyEnabled" = true;
+          "server.socksProxyHost" = "yourproxyhost.com";
+          "server.socksProxyPort" = "8080";
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.config.port ];
+
+    users.groups = mkIf (cfg.group == "tachidesk") {
+      tachidesk = { };
+    };
+
+    users.users = mkIf (cfg.user == "tachidesk") {
+      tachidesk = {
+        group = cfg.group;
+        home = cfg.dataDir;
+        description = "Tachidesk Daemon user";
+        isSystemUser = true;
+      };
+    };
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.dataDir}/.local/share/Tachidesk' 0700 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services.tachidesk-server =
+      let
+        #  HOCON is a JSON superset that tachidesk-server use for configuration
+        toHOCON = attr: (
+          let
+            attrType = builtins.typeOf attr;
+          in
+          if builtins.elem attrType [ "string" "path" "int" "float" ] then ''"${toString attr}"''
+          else if attrType == "bool" then lib.boolToString attr
+          # attrs, list, lambda, null
+          else
+            throw ''
+              [tachidesk-server]: invalid config value type '${attrType}'.
+            ''
+        );
+        configFile = pkgs.writeText "server.conf" (lib.strings.concatStrings [
+          ''
+            server.systemTrayEnabled = "${lib.boolToString cfg.enableSystemTray}"
+
+            server.ip = "${cfg.config.ip}"
+            server.port = "${toString cfg.config.port}"
+
+            server.downloadAsCbz = ${lib.boolToString cfg.config.downloadAsCbz}
+            server.localSourcePath = "${cfg.config.localSourcePath}"
+
+            server.basicAuthEnabled = ${lib.boolToString cfg.config.basicAuth.enable}
+          ''
+
+          (
+            if cfg.config.basicAuth.enable then
+              (if cfg.config.basicAuth.username == null || cfg.config.basicAuth.passwordFile == null then
+                throw ''
+                  [tachidesk-server]: the username and the password file cannot be null when the basic auth is enable
```

Comment:
I think these kinds of warnings are better suited for `assertions`. They aren't directly related to the configuration as much as it is a prerequisite for being able to configure the service.

```nix
assertions = [{
  assertion = cfg.config.basicAuth.enable -> (cfg.config.basicAuth.username != null && cfg.config.basicAuth.passwordFile != null);
  message = ''
    [tachidesk-server]: the username and the password file cannot be null when the basic auth is enable
  '';
}];
```

Reply:
I didn't know this syntax but I like it: I will surely reuse this. It's done